### DISCUSSION TITLE
How to debug Pyright LSP Server stdio communication?

### DISCUSSION THREAD

**Discussion by MarkBekooy** [microsoft/pyright#5945] [Q&A]
I'm creating an LSP client in Python to communicate with the Pyright LSP server. At the moment I start the Pyright server to communicate over stdio with the following code:

```python
p = subprocess.Popen(
	args=["pyright-langserver", "--stdio"],
	stdin=subprocess.PIPE,
	stdout=subprocess.PIPE,
	stderr=subprocess.PIPE,
)
```

Next, my client is instantiated from the `pylspclient` package and I'm sending successful `initialize` and `initialized` JSON-RPC messages, as these are printed to the console with their proper responses.

However, when something goes wrong, e.g. a misformatted `params` from my side, I don't really have a way to debug effectively...

Based on https://github.com/microsoft/pyright/blob/main/docs/build-debug.md#debugging-pyright, I got the debugger to work for both the extension and cli, but since I'm starting the Pyright LSP server in a subprocess, this way of debugging doesn't seem to work for me.

Any advice on getting the debugger to work would be very appreciated, because currently, I'm completely stuck.
(Even being able to see the JSON-RPC communication in an output channel would be beneficial)

**Comment by aspizu** [comment]
I'm trying to do the same thing
Here is my attempt which does **not** work for some reason.
```py
from __future__ import annotations
from pathlib import Path
from subprocess import PIPE, Popen
from typing import Any
import msgspec.json as json
from msgspec import Struct
from rich import print


class JsonRPC(Struct):
   jsonrpc: str
   method: str
   params: dict[str, Any]


proc = Popen(["pyright-langserver", "--stdio"], stdin=PIPE, stdout=PIPE)


def read():
   assert proc.stdout is not None
   content_length = int(
      proc.stdout.readline().removeprefix(b"Content-Length: ").removesuffix(b"\r\n")
   )
   return json.decode(proc.stdout.read(content_length + 2), type=JsonRPC)


def send(method: str, params: dict[str, Any]):
   assert proc.stdin is not None
   o = JsonRPC("2.0", method=method, params=params)
   content = json.encode(o)
   proc.stdin.write(
      b"Content-Length: " + str(len(content)).encode() + b"\r\n\r\n" + content
   )
   proc.stdin.flush()


print(read())
print(read())

send(
   "initialize",
   {
      "processId": None,
      "rootUri": None,
      "capabilities": {},
   },
)

print(read())

send(
   "textDocument/hover",
   {
      "textDocument": Path("a.py").absolute().as_uri(),
      "position": {
         "line": 0,
         "character": 0,
      },
   },
)

print(read())
```

**Comment by rchiodo** [comment]
You should be able to debug attach to the pyright node process if you pass it the `--inspect=<port>` argument when you start it. Here's an example blog post detailing how to debug node processes in VS code:

https://medium.com/@saransh98/debugging-node-js-via-vs-code-with-attach-single-process-only-87c50683a8ef
